### PR TITLE
Theme switching: Dynamic StatusBar and scroll bars

### PR DESCRIPTION
**Pull request by chrispader**
@grgia 

### Details

Currently the `StatusBar` and scrollbars across the whole app are not changing based on the theme. This PR adapts the `CustomStatusBar` component and adds a `CustomScrollbarWrapper` component.

### Fixed Issues

$ https://github.com/Expensify/App/issues/31915
PROPOSAL: 

### Tests

1. Add the following snippet to the render function of `Expensify.js`
```js
    useEffect(() => {
        let theme = 'dark';
        setInterval(() => {
            const newTheme = theme === 'dark' ? 'light' : 'dark';
            theme = newTheme;
            // eslint-disable-next-line rulesdir/prefer-actions-set-data
            Onyx.set(ONYXKEYS.PREFERRED_THEME, newTheme);
        }, 3000);
    }, []);
```
2. Log out/Go to "Sign in" page
3. Ensure that the status bar color (also UI of the browser) matches the dark theme of the page
4. Sign in
5. Theme will switch every 3 seconds.
6. Check if status bar content color is changing when theme switches (native and mWeb)
7. Check  if scrollbar color (e.g. in ReportScreen) is changing based on the active theme

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as in Tests.

### QA Steps

Same as in Tests.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot_20231129_091245_New Expensify Dev
Screenshot_20231129_091251_New Expensify Dev
Screenshot_20231129_091306_New Expensify Dev
Screenshot_20231129_091351_New Expensify Dev

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot_1701116313
Screenshot_1701116324
Screenshot_1701116336

</details>

<details>
<summary>iOS: Native</summary>

Simulator Screenshot - iPhone 15 Pro - 2023-11-27 at 21 35 13
Simulator Screenshot - iPhone 15 Pro - 2023-11-27 at 21 34 13
Simulator Screenshot - iPhone 15 Pro - 2023-11-27 at 21 34 15

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Simulator Screenshot - iPhone 15 Pro - 2023-11-27 at 21 14 18
Simulator Screenshot - iPhone 15 Pro - 2023-11-27 at 21 16 20
Simulator Screenshot - iPhone 15 Pro - 2023-11-27 at 21 16 23

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2023-11-27 at 21 21 51
Screenshot 2023-11-27 at 21 21 48
Screenshot 2023-11-27 at 21 19 59

</details>

<details>
<summary>MacOS: Desktop</summary>

Screenshot 2023-11-27 at 21 22 45

Screenshot 2023-11-27 at 21 23 16
Screenshot 2023-11-27 at 21 23 13

</details>

### CODE REVIEW

**Comment by chrispader** [comment]
cc @grgia

File: web/index.html [hunk 1]
```diff
@@ -31,7 +31,10 @@
         #root > div > div {
             height: 100% !important;
         }
-        :root {
+        .scrollbar-light {
+            color-scheme: light !important;
+        }
+        .scrollbar-light {
```
Comment (grgia) [inline web/index.html:37]: I think one of these should be scrollbar-dark no?

File: src/styles/styles.ts [hunk 2]
```diff
@@ -328,10 +328,6 @@ const styles = (theme: ThemeColors) =>
             textAlign: 'left',
         },
 
-        verticalAlignMiddle: {
-            verticalAlign: 'middle',
-        },
-
```
Comment (grgia) [inline src/styles/styles.ts:334]: is this unused?

File: src/styles/StyleUtils.ts [hunk 3]
```diff
@@ -1390,6 +1390,19 @@ function getDotIndicatorTextStyles(isErrorText = true): TextStyle {
     return isErrorText ? {...styles.offlineFeedback.text, color: styles.formError.color} : {...styles.offlineFeedback.text};
 }
 
+/**
+ * Returns container styles for showing the icons in MultipleAvatars/SubscriptAvatar
+ */
```
Comment (grgia) [inline src/styles/StyleUtils.ts:1395]: comment update?

File: src/components/CustomStatusBar/index.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import CustomStatusBarContext from './CustomStatusBarContext';
+
+type CustomStatusBarProps = {
+    isNested: boolean;
+};
+
+function CustomStatusBar({isNested = false}: CustomStatusBarProps): React.ReactElement | null {
+    const {isRootStatusBarDisabled, disableRootStatusBar} = useContext(CustomStatusBarContext);
+    const theme = useTheme();
+    const statusBarContentTheme = useMemo(() => (theme.statusBarContentTheme === 'light' ? 'light-content' : 'dark-content'), [theme.statusBarContentTheme]);
```
Comment (grgia) [inline src/components/CustomStatusBar/index.tsx:14]: use CONST for these

File: src/components/CustomScrollbarWrapper/index.tsx [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import {View} from 'react-native';
+import Themes from '@styles/themes/Themes';
+import useThemePreferenceWithStaticOverride from '@styles/themes/useThemePreferenceWithStaticOverride';
+import CustomScrollbarWrapperProps from './CustomScrollbarWrapperProps';
+
+function CustomScrollbarWrapper({children, theme: staticThemePreference}: CustomScrollbarWrapperProps): React.ReactElement {
+    const preferredTheme = useThemePreferenceWithStaticOverride(staticThemePreference);
+    const scrollbarTheme = useMemo<CSSProperties['colorScheme']>(() => Themes[preferredTheme].scrollBarTheme, [preferredTheme]);
+
+    return <View style={{flex: 1, colorScheme: scrollbarTheme}}>{children}</View>;
```
Comment (grgia) [inline src/components/CustomScrollbarWrapper/index.tsx:11]: could we avoid inline styles for this?

File: web/index.html (hunk 1 above)
Comment (chrispader) [reply web/index.html:37]: actually, forgot to remove this. This is obsolete

**Comment by chrispader** [comment]
I also added the functionality to show different status bar content styles (`light-content`/`dark-content`) on different screens, by passing `statusBarStyle` to the `theme.PAGE_THEMES` property

**Comment by chrispader** [comment]
Just merged `main`, not sure if this fixed it. Haven't experienced this error...

**Comment by grgia** [comment]
@fedirjh are you able to prioritize this one?

**Comment by chrispader** [comment]
> I just noticed this bug, the header style is updated before the navigation, this is not repo on `main`.

That's not a bug, i'm currently just animating the transition of the scrollbar. I tried not animating it, but then sometimes the scrollbar would be updated earlier/later than the content, which looks off. I think animating it is the best approach

**Comment by chrispader** [comment]
> @chrispader @grgia Should the scrollbar style be opposite to the theme? For example, if the theme is dark, scrollbar should be light and vice versa? 

No the scrollbar style should be the same as the theme, since the scrollbar's container should (almost) match the background

**Review by fedirjh — COMMENTED** [review]
Bug: mWeb Chrome status bar is not updated with dark scheme.

| Production | PR |
|--------|--------|
| Screenshot 2023-11-30 at 1 26 30 AM | Screenshot 2023-11-30 at 1 34 12 AM | 


---

Bug : Sometimes I get this bug, I just try to open settings modal:

Screenshot 2023-11-30 at 12 51 23 AM

https://github.com/Expensify/App/assets/36869046/570efd0d-40dc-4672-b9a6-81a525f31041

**Comment by chrispader** [comment]
> Bug: mWeb Chrome status bar is not updated with dark scheme.

I'm not sure if this is actually a problem with this PR. When i serve `main` with the `webpack-dev-server`, the header bar color of Chrome also doesn't change. I suppose this somehow has to do with viewing an unsafe `https` page?

`main` via `npm run web`:
Screenshot_1701342281

**Comment by chrispader** [comment]
> Bug : Sometimes I get this bug, I just try to open settings modal:

Ok, so this is not related to this pR

> The route bug is being discussed here btw - https://expensify.slack.com/archives/C01GTK53T8Q/p1701334362644509

**Comment by chrispader** [comment]
@grgia could you please create AdHoc builds for this PR? :)

File: ios/NewExpensifyTests/Info.plist [hunk 6]
```diff
@@ -15,10 +15,10 @@
 	<key>CFBundlePackageType</key>
 	<string>BNDL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>1.4.5</string>
+	<string>1.4.6</string>
```
Comment (grgia) [inline ios/NewExpensifyTests/Info.plist:18]: Should we commit these?

**Review by grgia — DISMISSED** [review]

**Review by fedirjh — COMMENTED** [review]
We have some lint errors.

**Review by fedirjh — APPROVED** [review]
Looks good to me.

**Review by grgia — APPROVED** [review]

**Comment by kbecciv** [comment]
@chrispader @grgia QA team is seeing the console error when adding snippet
image

**Comment by grgia** [comment]
@kbecciv apologies, that should not be included in QA steps.
The QA is regression testing for status bar colors. I can internally QA

File: src/components/CustomStatusBar/index.tsx [hunk 7]
```diff
@@ -1,29 +1,91 @@
...
 import StatusBar from '@libs/StatusBar';
 import useTheme from '@styles/themes/useTheme';
-import type CustomStatusBarType from './types';
+import CustomStatusBarContext from './CustomStatusBarContext';
+
+type CustomStatusBarProps = {
+    isNested: boolean;
+};
+
+const propTypes = {
```
Comment (blazejkustra) [inline src/components/CustomStatusBar/index.tsx:13]: @chrispader Adding Prop types in TS files is against our TS guidelines, can you make a follow up PR that removes these?

File after changes should look more or less like this:
```ts
type CustomStatusBarProps = {
    /** Whether the CustomStatusBar is nested within another CustomStatusBar.
     *  A nested CustomStatusBar will disable the "root" CustomStatusBar. */
    isNested: boolean;
};

function CustomStatusBar({isNested = false}: CustomStatusBarProps) {
    const {isRootStatusBarDisabled, disableRootStatusBar} = useContext(CustomStatusBarContext);
    const theme = useTheme();

    const isDisabled = !isNested && isRootStatusBarDisabled;

    useEffect(() => {
        if (isNested) {
            disableRootStatusBar(true);
        }

        return () => {
            if (!isNested) {
                return;
            }
            disableRootStatusBar(false);
        };
    }, [disableRootStatusBar, isNested]);

    const updateStatusBarStyle = useCallback<EventListenerCallback<NavigationContainerEventMap, 'state'>>(() => {
        if (isDisabled) {
            return;
        }

        // Set the status bar colour depending on the current route.
        // If we don't have any colour defined for a route, fall back to
        // appBG color.
        const currentRoute = navigationRef.getCurrentRoute();

        let currentScreenBackgroundColor = theme.appBG;
        let statusBarStyle = theme.statusBarStyle;
        if (currentRoute && 'name' in currentRoute && currentRoute.name in theme.PAGE_THEMES) {
            const screenTheme = theme.PAGE_THEMES[currentRoute.name];
            currentScreenBackgroundColor = screenTheme.backgroundColor;
            statusBarStyle = screenTheme.statusBarStyle;
        }

        updateStatusBarAppearance({backgroundColor: currentScreenBackgroundColor, statusBarStyle});
    }, [isDisabled, theme.PAGE_THEMES, theme.appBG, theme.statusBarStyle]);

    useEffect(() => {
        navigationRef.addListener('state', updateStatusBarStyle);

        return () => navigationRef.removeListener('state', updateStatusBarStyle);
    }, [updateStatusBarStyle]);

    useEffect(() => {
        if (isDisabled) {
            return;
        }

        updateStatusBarAppearance({statusBarStyle: theme.statusBarStyle});
    }, [isDisabled, theme.statusBarStyle]);

    if (isDisabled) {
        return null;
    }

    return <StatusBar />;
}

CustomStatusBar.displayName = 'CustomStatusBar';

export default CustomStatusBar;
```

cc @fabioh8010

File: src/components/CustomStatusBar/index.tsx [hunk 8]
```diff
@@ -1,29 +1,91 @@
...
+
+        let currentScreenBackgroundColor = theme.appBG;
+        let statusBarStyle = theme.statusBarStyle;
+        if (currentRoute && 'name' in currentRoute && currentRoute.name in theme.PAGE_THEMES) {
+            const screenTheme = theme.PAGE_THEMES[currentRoute.name];
+            currentScreenBackgroundColor = screenTheme.backgroundColor;
+            statusBarStyle = screenTheme.statusBarStyle;
+        }
+
+        StatusBar.setBackgroundColor(currentScreenBackgroundColor, true);
```
Comment (roryabraham) [inline src/components/CustomStatusBar/index.tsx:63]: @chrispader I notice that you removed the animated transition between status bar background colors that was present in `NavigationRoot.tsx`. Was there any discussion around that choice? Can you add it back here please?

Comment (chrispader) [reply src/components/CustomStatusBar/index.tsx:63]: Yes, sorry my mistake. Just added a PR here: https://github.com/Expensify/App/pull/33200

**Pull request merged**